.github/workflows/build.yaml

```diff
@@ -0,0 +1,121 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+on:
+  workflow_call:
+    inputs:
+      build-matrix:
+        description: JSON string of charms to build
+        required: true
+        type: string
+        # Example:
+        # [
+        #   {
+        #     "job_display_name": "Build mysql charm | focal",
+        #     "bases_index": 0, # corresponds to charmcraft.yaml file
+        #     "path": "."
+        #   },
+        #   {
+        #     "job_display_name": "Build mysql charm | jammy",
+        #     "bases_index": 1,
+        #     "path": "."
+        #   },
+        #   {
+        #     "job_display_name": "Build application charm | jammy",
+        #     "bases_index": 0,
+        #     "path": "tests/integration/relations/application-charm"
+        #   }
+        # ]
+
+jobs:
+  build:
+    strategy:
+      matrix:
+        charm: ${{ fromJSON(inputs.build-matrix) }}
+    name: ${{ matrix.charm.job_display_name }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Setup environment
+        # TODO: Replace with custom image on self-hosted runner
+        run: |
+          # Copied from https://github.com/charmed-kubernetes/actions-operator/blob/96fb0b07eb675f74cf1796be812bc7e67a0d62fc/src/bootstrap/index.ts#L151
+          sudo adduser $USER lxd
+          newgrp lxd
+          sudo lxd waitready
+          sudo lxd init --auto
+          sudo iptables -F FORWARD
+          sudo iptables -P FORWARD ACCEPT
+          sudo chmod a+wr /var/snap/lxd/common/lxd/unix.socket
+          sudo snap install charmcraft --classic
+      - name: Get charmcraft version
+        id: charmcraft-version
+        # Major and minor versions (e.g. "2.1")
+        run: echo "version=$(charmcraft version | cut --delimiter '.' --fields 1,2)" >> $GITHUB_OUTPUT
+      - name: Restore cache of `charmcraft pack` LXC instance
+        id: restore-cache
+        uses: actions/cache/restore@v3
+        with:
+          path: ~/ga-charmcraft-cache/**
+          key: charmcraft-pack-${{ matrix.charm.path }}-${{ matrix.charm.bases_index }}-${{ steps.charmcraft-version.outputs.version }}-${{ hashFiles(format('{0}/charmcraft.yaml', matrix.charm.path), format('{0}/actions.yaml', matrix.charm.path), format('{0}/requirements.txt', matrix.charm.path)) }}
+      - name: Import cached container
```

Comment:
to confirm: if the CI is scheduled (nightly), we will import the cached LXD container for `main`, run charmcraft pack, delete cached LXD container for `main`, and then save the LXD container from the charmcraft pack step?

if so, i have the following suggestion: we should not import any cached LXD containers if scheduled, run charmcraft pack, then save the LXD container from the charmcraft pack step. This will result in a fresh LXD container to be created and cached